### PR TITLE
Changes necessary to support `netctl login`

### PR DESCRIPTION
**Pull request by dseevr**
These are the changes from https://github.com/contiv/modelgen/pull/24.  Regular commit message follows:

This commit does not break any existing functionality for regular netmaster communication

Changes:

- default http client is replaced by a http.Client in the ContivClient struct
- httpGet()/httpPost()/httpDelete() are moved into ContivClient and use the new internal http.Client
- added SetHttpClient() which allows for overriding the internal http.Client (to disable cert checking, add timeouts, etc.)
- added Login() function which performs a login POST request against auth_proxy
- added SetAuthToken() which adds a custom request header (X-Auth-Token) before requests are sent out
- httpGet()/httpPost()/httpDelete() are modified to set any custom request headers before performing the request

Signed-off-by: Bill Robinson <dseevr@users.noreply.github.com>

### CODE REVIEW

File: client/contivModelClient.go [hunk 1]
```diff
@@ -80,7 +136,16 @@ type Link struct {
 
 func httpGet(url string, jdata interface{}) error {
 
-	r, err := http.Get(url)
+	req, err := http.NewRequest("GET", url, nil)
+	if err != nil {
+		return err
+	}
+
+	if err := addAuthTokenHeader(req); err != nil {
```
Comment (yuva29) [inline client/contivModelClient.go]: Are we enforcing that all the requests should go through proxy? What if the user wants to bypass proxy and make a call to netmaster directly?

Comment (dseevr) [reply client/contivModelClient.go]: See line 95.  Non-HTTPS requests return early and don't get the header added (for security reasons) so this doesn't interfere with normal operation

Non-TLS netmasters can still be talked to normally (omit `--netmaster` for the default on `localhost:9999` or specify one manually)

Comment (yuva29) [reply client/contivModelClient.go]: Got it.

**Review by rhim — APPROVED** [review]
LGTM.

File: client/contivModelClient.go [hunk 2]
```diff
@@ -64,14 +64,70 @@ package client
...
+	tr := &http.Transport{
+		TLSClientConfig: &tls.Config{InsecureSkipVerify: true},
+	}
+
+	httpClient = &http.Client{Transport: tr}
+}
+
+type NetctlConfig struct {
+	Token string `json:"token"`
+}
```
Comment (jojimt) [inline client/contivModelClient.go]: Let's separate out the netctl specific items and move them out of the model as discussed in the modelgen PR.

**Review by jojimt — CHANGES_REQUESTED** [review]

**Review by rhim — COMMENTED** [review]
why are there changes for check package in this patch?

**Comment by dseevr** [comment]
@rhim that was from bringing in contiv's fork of gocheck.  I broke that out into a separate commit

**Review by rhim — APPROVED** [review]
These changes look fine. I am a bit unclear on tests that we need for auth_proxy - basically test that ensures that no command can be run without token.

Another test would be that of invalid/expired token. 

As long as we have those scenarios covered, I think this is good to go.

Please also open the doc change PR for this change.

**Comment by dseevr** [comment]
@rhim there's no problem with netctl sending requests to auth_proxy without a token (e.g., if they haven't logged in yet or deleted the config after logging in) because they will get back a 401 unauthorized.  auth_proxy has tests to cover that scenario.  It also has tests to cover invalid and expired tokens.

Do you want netctl to refuse to send the requests if the target is https and no token (i.e., config) is present?

**Comment by rhim** [comment]
No, that’s ok. Shouldn’t overburden client if server side mechanism already exists.

- Himanshu
     .:|:.:|:.

On May 4, 2017, at 5:29 PM, Bill Robinson <notifications@github.com<mailto:notifications@github.com>> wrote:


@rhim<https://github.com/rhim> there's no problem with netctl sending requests to auth_proxy without a token (e.g., if they haven't logged in yet or deleted the config after logging in) because they will get back a 401 unauthorized. auth_proxy has tests to cover that scenario. It also has tests to cover invalid and expired tokens.

Do you want netctl to refuse to send the requests if the target is https and no token (i.e., config) is present?

**Pull request merged**